lib/common/socket.c

```diff
@@ -999,6 +1044,17 @@ void h2o_socket_ssl_async_resumption_setup_ctx(SSL_CTX *ctx)
     /* if necessary, it is the responsibility of the caller to disable the internal cache */
 }
 
+static void on_session_cache_entry_destroy(h2o_iovec_t value)
+{
+    SSL_SESSION *session = (SSL_SESSION *)value.base;
+    SSL_SESSION_free(session);
+}
+
+h2o_cache_t *h2o_socket_ssl_create_session_cache(int flags, size_t capacity, uint64_t duration)
+{
+    return h2o_cache_create(flags, capacity, duration, on_session_cache_entry_destroy);
+}
```

Comment:
If we are to provide abstraction of the cache (i.e. this create function), then the type of the cache should be a type that wraps `h2o_cache_t`, instead of `h2o_cache_t` itself.

Instead, if we are to consider the type of the cache to be `h2o_cache_t`, the dispose callback should be exposed as a global function, and the user of the socket implementation should (be able to) construct the cache without using an abstraction.

For simplicity, my preference goes to the latter.